[mono] Implement synch block fast paths in managed

lambdageek opened the pull request:
Investigate one incremental alternative from https://github.com/dotnet/runtime/issues/48058

Rather than switching to NativeAOT's Monitor implementation in one step, first get access to Mono's synchronization block from managed, and implement the fast paths for hash code and monitor code in C#.

lambdageek committed:
[mono] Implement synch block fast paths in managed

Investigate one incremental alternative from https://github.com/dotnet/runtime/issues/48058

Rather than switching to NativeAOT's Monitor implementation in one
step, first get access to Mono's synchronization block from managed,
and implement the fast paths for hash code and monitor code in C#.

lambdageek committed:
Implement TryEnterFast and TryEnterInflatedFast

This should help the interpreter particularly, since it doesn't have
an intrinsic path here

lambdageek committed:
wip

lambdageek committed:
Add replacements for test_owner and test_synchronized icalls

ObjectHeader.IsEntered and ObjectHeader.HasOwner, respectively

vargaz commented:
This might lead to perf problems since the previous optimized c code is replaced by the not so optimized JITted code.

lambdageek commented:
> This might lead to perf problems since the previous optimized c code is replaced by the not so optimized JITted code.

@vargaz not everything has intrinsics (Monitor_test_synchronised and Monitor_test_owner were just plain icalls).  Also the interpreter doesn't have intrinsics for `Enter(object, bool&)`.  

On JIT/AOT the intrinsic is:
```c
  if (mono_monitor_enter_v4_fast (_lock, ref lockTaken)) // no wrapper icall
    goto done;
 mono_monitor_enter_v4_internal (_lock, ref lockTaken); // icall with wrapper
 done:
```

I think it probably makes sense to keep this, but it's also possible that `mono_monitor_enter_v4_fast` is slower than a managed fast path. With aggressive inlining it could just turn into a volatile load and a couple of ALU ops and a CompareExchange.  That might be cheaper than an icall.

For interp+AOT I think it will for sure be cheaper than a interp->native transition.

lambdageek committed:
use sequential layout for LockWord

lambdageek committed:
remove managed Inflate (ifdef'd out)

it needs at least alloc_mon/discard_mon icalls which might kill any
advantage of having a purely managed path

lambdageek committed:
access sync block through helpers

lambdageek committed:
add DumpRef to print out pointer values from ObjectHeader

lambdageek committed:
Get a pointer to the object header another way

Unsafe.As<TFrom,TTo>() was converting a object** into a header*.

Go through Unsafe.AsPointer to get a Header** and then deref it once
to get a header

lambdageek committed:
remove DumpRef

lambdageek committed:
XXX temporary return from MonoResolveUnmanagedDll if context is null

There's some kind of situation where apparently Default is seen to be
null - so we got into a loop between the managed hash code
implementation and the static constructor of AssemblyLoadContext

lambdageek committed:
XXX some bug in GetHashCode :-(

lambdageek committed:
BUGFIX: logical shifts, not arithmetic

on wasm32, IntPtr and int are both the same size. so when we store a
hash code into _lock_word that ends up with the 30th high bit set, the
right shift to get it back out needs to be arithmetic or else we get
incorrect hash codes

lambdageek committed:
XXX disable hacks

lambdageek committed:
put back the ALC hack

lambdageek committed:
force inlining

lambdageek committed:
move fast path into ReliableEnterTimeout

lambdageek committed:
Don't pass lockTaken to ObjectHolder.TryEnterFast

set it in the caller if we succeeded

lambdageek committed:
TryEnterFast: handle flat recursive locking, too

lambdageek committed:
whitespace

lambdageek committed:
Implement Monitor.Exit fast path in managed

If the LockWord is flat, there has been no contention, so there is
noone to signal when we exit

lambdageek committed:
Don't call MonoResolveUnmanagedDll for the default ALC

it would always return null anyway.

But also this avoids a circular dependency between the managed Default
ALC constructor and Monitor icalls (the ALC base constructor locks
allContexts).

lambdageek committed:
[interp] don't lookup internal calls that were replaced by an opcode

If it was replaced by an intrinsic, we're not really going to call the
wrapper, no need to look for it

lambdageek commented:
On one machine, I see this from running WASM System.Runtime.Tests using v8 without AOT:

Baseline:
```
  info: Total: 50616, Errors: 0, Failed: 0, Skipped: 92, Time: 130.155189s
```

This PR:
```
  info: Total: 50616, Errors: 0, Failed: 0, Skipped: 92, Time: 130.30528s
```

I'm planning to look at optimizing the interpreter IR.  For single threaded wasm every `lock(obj)` should be able to avoid going into native code except if `obj` has its hash code taken and was locked

lambdageek commented:
Interesting. I didn't realize it, but we don't have `Interlocked.CompareExchange` and `Volatile.Read` intrinsics in the interpreter. That's unfortunate for this project.

lambdageek committed:
Add Interlocked.CompareExchange interp intrinsics for I4 and I8

lambdageek committed:
ask EE to inline LockWordCompareExchange

lambdageek commented:
Added CompareExchange and some `AggressiveInlining` hints.  I now see this for the optimized IR for `ReliableEnterTimeout` on amd64 desktop. As far as I can tell, the fast path has calls to `ThrowIfNull`, `GC.KeepAlive`, `Thread.CurrentThread` and `Thread.GetSmallId`.  After that we're manipulating the LockWord directly.  I'll get updated timing for System.Runtime.Tests shortly

```
Runtime method: System.Threading.Monitor:ReliableEnterTimeout (object,int,bool&) 0x16da01470
Locals size 128
Calculated stack height: 4, stated height: 8
IR_0000: ldstr          [104 <- nil], 0
IR_0003: mov.8          [96 <- 0],
IR_0006: call           [48 <- 96], 1
IR_000a: bge.i4.imm.sp  [nil <- 8], 0, IR_001c
IR_000e: beq.i4.imm.sp  [nil <- 8], 65535, IR_001c
IR_0012: ldstr          [104 <- nil], 2
IR_0015: newobj         [48 <- 96], 4
IR_001a: throw          [nil <- 48],
IR_001c: mov.8          [40 <- 0],
IR_001f: ldfld.i8       [48 <- 0], 8
IR_0023: mov.8          [48 <- 48],
IR_0026: mov.8          [96 <- 0],
IR_0029: call           [56 <- 96], 5
IR_002d: mov.vt         [32 <- 48], 8
IR_0031: mov.8          [48 <- 32],
IR_0034: bne.un.i8.imm.sp [nil <- 48], 0, IR_0070
IR_0038: call           [48 <- 96], 6
IR_003c: cknull         [96 <- 48],
IR_003f: call           [48 <- 96], 7
IR_0043: conv.i8.u4     [48 <- 48],
IR_0046: shl.i8.imm     [48 <- 48], 10
IR_004a: mov.8          [48 <- 48],
IR_004d: mov.vt         [56 <- 32], 8
IR_0051: ldflda         [64 <- 40], 8
IR_0055: mov.8          [72 <- 48],
IR_0058: mov.8          [80 <- 56],
IR_005b: mono_interlocked.cmpxchg.i8 [48 <- 64 72 80],
IR_0060: mov.8          [96 <- 40],
IR_0063: call           [56 <- 96], 5
IR_0067: mov.8          [56 <- 32],
IR_006a: bne.un.i8.s    [nil <- 48 56], IR_00f1
IR_006e: br.s           [nil <- nil], IR_00eb
IR_0070: mov.8          [48 <- 32],
IR_0073: ldc.i8         [56 <- nil], 2
IR_0079: and.i8         [48 <- 48 56],
IR_007d: ble.un.i8.imm.sp [nil <- 48], 0, IR_008d
IR_0081: mov.8          [96 <- 40],
IR_0084: call           [48 <- 96], 8
IR_0088: brfalse.i4.s   [nil <- 48], IR_00f1
IR_008b: br.s           [nil <- nil], IR_00eb
IR_008d: mov.8          [48 <- 32],
IR_0090: ldc.i8         [56 <- nil], 3
IR_0096: and.i8         [48 <- 48 56],
IR_009a: bne.un.i8.imm.sp [nil <- 48], 0, IR_00f1
IR_009e: call           [48 <- 96], 6
IR_00a2: cknull         [96 <- 48],
IR_00a5: call           [48 <- 96], 7
IR_00a9: mov.8          [56 <- 32],
IR_00ac: shr.un.i8.imm  [56 <- 56], 10
IR_00b0: bne.un.i4.s    [nil <- 56 48], IR_00f1
IR_00b4: ldloca.s       [96 <- nil], 32
IR_00b7: call           [48 <- 96], 9
IR_00bb: brfalse.i4.s   [nil <- 48], IR_00c0
IR_00be: br.s           [nil <- nil], IR_00f1
IR_00c0: mov.8          [48 <- 32],
IR_00c3: add.i8.imm     [48 <- 48], 4
IR_00c7: mov.8          [48 <- 48],
IR_00ca: mov.vt         [56 <- 32], 8
IR_00ce: ldflda         [64 <- 40], 8
IR_00d2: mov.8          [72 <- 48],
IR_00d5: mov.8          [80 <- 56],
IR_00d8: mono_interlocked.cmpxchg.i8 [48 <- 64 72 80],
IR_00dd: mov.8          [96 <- 40],
IR_00e0: call           [56 <- 96], 5
IR_00e4: mov.8          [56 <- 32],
IR_00e7: bne.un.i8.s    [nil <- 48 56], IR_00f1
IR_00eb: ldc.i4.1       [48 <- nil],
IR_00ed: stind.i1       [nil <- 16 48],
IR_00f0: ret.void       [nil <- nil],
IR_00f1: ldc.i4.1       [112 <- nil],
IR_00f3: mov.8.3        [nil <- nil], 96 <- 0, 104 <- 8, 120 <- 16
IR_00fa: call           [48 <- 96], 10
IR_00fe: ret.void       [nil <- nil],
```

lambdageek commented:
With intrinsics

```
  info: Total: 50616, Errors: 0, Failed: 0, Skipped: 92, Time: 129.2762049s
```

lambdageek committed:
Use a ref struct to pass around the address of a MonoObject* on the stack

This is similar to ObjectHandleOnStack except with a getter that lets
us view the object header.

Get rid of calls to GC.KeepAlive.

With this, the fast path (locking a flat unowned object) doesn't have
any calls on it besides argument validation

lambdageek commented:
One final update.   I think this is ready for review @vargaz @kg @BrzVlad 

Used a `ObjectHeaderOnStack` ref struct to pass around the `ref Header` that we need while also keeping the underlying object pointer on the stack.  With this the two fast path (locking an unlocked object - either flat or previously inflated) only has calls for argument validation and is otherwise just ALU ops and a CompareExchange intrinsic.

<details>
<summary>Interpreter IR for ReliableEnterTimeout</summary>
<pre>
Runtime method: System.Threading.Monitor:ReliableEnterTimeout (object,int,bool&) 0x16b9d5470
Locals size 144
Calculated stack height: 4, stated height: 8
IR_0000: ldstr          [120 <- nil], 0
IR_0003: mov.8          [112 <- 0],
IR_0006: call           [64 <- 112], 1
IR_000a: bge.i4.imm.sp  [nil <- 8], 0, IR_001c
IR_000e: beq.i4.imm.sp  [nil <- 8], 65535, IR_001c
IR_0012: ldstr          [120 <- nil], 2
IR_0015: newobj         [64 <- 112], 4
IR_001a: throw          [nil <- 64],
IR_001c: mov.8          [32 <- 0],
IR_001f: ldloca.s       [64 <- nil], 32
IR_0022: newobj_vt_inlined [80 <- 72], 8
IR_0026: stfld.i8       [nil <- 80 64], 0
IR_002a: mov.vt         [48 <- 72], 8
IR_002e: mov.8          [64 <- 72],
IR_0031: ldind.i8       [64 <- 64],
IR_0034: ldfld.i8       [64 <- 64], 8
IR_0038: mov.8          [64 <- 64],
IR_003b: mov.vt         [40 <- 64], 8
IR_003f: mov.8          [64 <- 40],
IR_0042: bne.un.i8.imm.sp [nil <- 64], 0, IR_007d
IR_0046: call           [64 <- 112], 5
IR_004a: cknull         [112 <- 64],
IR_004d: call           [64 <- 112], 6
IR_0051: conv.i8.u4     [64 <- 64],
IR_0054: shl.i8.imm     [64 <- 64], 10
IR_0058: mov.8          [64 <- 64],
IR_005b: mov.vt         [72 <- 40], 8
IR_005f: mov.8          [80 <- 48],
IR_0062: ldind.i8       [80 <- 80],
IR_0065: ldflda         [80 <- 80], 8
IR_0069: mov.8          [88 <- 64],
IR_006c: mov.8          [96 <- 72],
IR_006f: mono_interlocked.cmpxchg.i8 [64 <- 80 88 96],
IR_0074: mov.8          [72 <- 40],
IR_0077: bne.un.i8.s    [nil <- 64 72], IR_00fe
IR_007b: br.s           [nil <- nil], IR_00f8
IR_007d: mov.8          [64 <- 40],
IR_0080: ldc.i8         [72 <- nil], 2
IR_0086: and.i8         [64 <- 64 72],
IR_008a: ble.un.i8.imm.sp [nil <- 64], 0, IR_009b
IR_008e: mov.vt         [112 <- 48], 8
IR_0092: call           [64 <- 112], 7
IR_0096: brfalse.i4.s   [nil <- 64], IR_00fe
IR_0099: br.s           [nil <- nil], IR_00f8
IR_009b: mov.8          [64 <- 40],
IR_009e: ldc.i8         [72 <- nil], 3
IR_00a4: and.i8         [64 <- 64 72],
IR_00a8: bne.un.i8.imm.sp [nil <- 64], 0, IR_00fe
IR_00ac: call           [64 <- 112], 5
IR_00b0: cknull         [112 <- 64],
IR_00b3: call           [64 <- 112], 6
IR_00b7: mov.8          [72 <- 40],
IR_00ba: shr.un.i8.imm  [72 <- 72], 10
IR_00be: bne.un.i4.s    [nil <- 72 64], IR_00fe
IR_00c2: ldloca.s       [112 <- nil], 40
IR_00c5: call           [64 <- 112], 8
IR_00c9: brfalse.i4.s   [nil <- 64], IR_00ce
IR_00cc: br.s           [nil <- nil], IR_00fe
IR_00ce: mov.8          [64 <- 40],
IR_00d1: add.i8.imm     [64 <- 64], 4
IR_00d5: mov.8          [64 <- 64],
IR_00d8: mov.vt         [72 <- 40], 8
IR_00dc: mov.8          [80 <- 48],
IR_00df: ldind.i8       [80 <- 80],
IR_00e2: ldflda         [80 <- 80], 8
IR_00e6: mov.8          [88 <- 64],
IR_00e9: mov.8          [96 <- 72],
IR_00ec: mono_interlocked.cmpxchg.i8 [64 <- 80 88 96],
IR_00f1: mov.8          [72 <- 40],
IR_00f4: bne.un.i8.s    [nil <- 64 72], IR_00fe
IR_00f8: ldc.i4.1       [64 <- nil],
IR_00fa: stind.i1       [nil <- 16 64],
IR_00fd: ret.void       [nil <- nil],
IR_00fe: ldc.i4.1       [128 <- nil],
IR_0100: mov.8.3        [nil <- nil], 112 <- 0, 120 <- 8, 136 <- 16
IR_0107: call           [64 <- 112], 9
IR_010b: ret.void       [nil <- nil],
</pre>
</summary>

BrzVlad commented:
The numbers from the library tests suite are not really useful, since there is no significant difference there. I was able to miraculously find the benchmark suite that I used for testing when I implemented the monitor thin locks ages ago. Running those benchmarks on jit and interp, with and without your change might provide some better context of the performance implications of this.
[tests.zip](https://github.com/dotnet/runtime/files/10577280/tests.zip)

kg commented:
> only has calls for argument validation and is otherwise just ALU ops and a CompareExchange intrinsic.

Can the argument validation methods become aggressive inlined so that the fast path never performs a call? Calls aren't cheap in interp afaik, those alone might be worse than doing all this in an icall. You could put the failure case for validation in a call

lambdageek committed:
Add fast path for Monitor.Exit for inflated monitors with no waiters

lambdageek committed:
Check for obj == null before ThrowIfNull

In the interp it's cheaper to do a null check followed by a call than
an unconditional call

lambdageek committed:
Inline RuntimeHelpers.GetHashCode

lambdageek commented:
I made a [little benchmark](https://gist.github.com/lambdageek/c53949741489cac862d7918f2132b5a9) that mostly tests the uncontested flat & `lock (obj) {}` and `obj.GetHashCode()`.  I compared using BenchmarkDotNet against the base commit of this PR.

The **miniJIT results** are better or roughly the same for each microbenchmark

|                       Method |        Job |                                                                         Toolchain |      Mean |     Error |    StdDev | Ratio |
|----------------------------- |----------- |---------------------------------------------------------------------------------- |----------:|----------:|----------:|------:|
| LockPulseUnlockFlatObjNoHash | Job-UTZHRH | baseline | 41.762 ns | 0.4819 ns | 0.4508 ns |  1.00 |
| LockPulseUnlockFlatObjNoHash | Job-BGNBMX |        PR | 34.284 ns | 0.2759 ns | 0.2580 ns |  0.82 |
|                              |            |                                                                                   |           |           |           |       |
|      LockUnlockFlatObjNoHash | Job-UTZHRH | baseline | 21.351 ns | 0.1881 ns | 0.1759 ns |  1.00 |
|      LockUnlockFlatObjNoHash | Job-BGNBMX |        PR | 20.517 ns | 0.0568 ns | 0.0532 ns |  0.96 |
|                              |            |                                                                                   |           |           |           |       |
|            GetHashCodeNoLock | Job-UTZHRH | baseline |  9.715 ns | 0.0884 ns | 0.0826 ns |  1.00 |
|            GetHashCodeNoLock | Job-BGNBMX |        PR |  4.814 ns | 0.0454 ns | 0.0424 ns |  0.50 |
|                              |            |                                                                                   |           |           |           |       |
|     LockUnlockInflatedObject | Job-UTZHRH | baseline | 23.887 ns | 0.2497 ns | 0.2335 ns |  1.00 |
|     LockUnlockInflatedObject | Job-BGNBMX |        PR | 23.474 ns | 0.0305 ns | 0.0270 ns |  0.98 |

The **interpreter** results (with `MONO_ENV_OPTIONS=--interp=-tiering`) are mostly the same or better except for `GetHashCode` which pretty bad and lock/unlock for an inflated object (object which previously had contention or had its hash code retrieved) which is about 20% slower

|                       Method |        Job |                                                                         Toolchain |      Mean |    Error |   StdDev | Ratio | RatioSD |
|----------------------------- |----------- |---------------------------------------------------------------------------------- |----------:|---------:|---------:|------:|--------:|
| LockPulseUnlockFlatObjNoHash | Job-XWWCLM | baseline | 211.45 ns | 2.035 ns | 1.904 ns |  1.00 |    0.00 |
| LockPulseUnlockFlatObjNoHash | Job-KJZRXO |        PR | 180.90 ns | 1.664 ns | 1.556 ns |  0.86 |    0.01 |
|                              |            |                                                                                   |           |          |          |       |         |
|      LockUnlockFlatObjNoHash | Job-XWWCLM | baseline | 111.35 ns | 1.321 ns | 1.236 ns |  1.00 |    0.00 |
|      LockUnlockFlatObjNoHash | Job-KJZRXO |        PR | 101.84 ns | 0.930 ns | 0.870 ns |  0.91 |    0.02 |
|                              |            |                                                                                   |           |          |          |       |         |
|            GetHashCodeNoLock | Job-XWWCLM | baseline |  18.46 ns | 0.242 ns | 0.227 ns |  1.00 |    0.00 |
|            GetHashCodeNoLock | Job-KJZRXO |        PR |  38.81 ns | 0.419 ns | 0.392 ns |  2.10 |    0.04 |
|                              |            |                                                                                   |           |          |          |       |         |
|     LockUnlockInflatedObject | Job-XWWCLM | baseline | 113.51 ns | 1.102 ns | 1.031 ns |  1.00 |    0.00 |
|     LockUnlockInflatedObject | Job-KJZRXO |        PR | 137.32 ns | 0.446 ns | 0.417 ns |  1.21 |    0.01 |

lambdageek committed:
inline TryEnterInflatedFast; nano-opts

- change SyncBlock.HashCode to return the value, not a ref - we don't
have managed code that writes into the sync block.
- change TryEnterInflatedFast to avoid looping - if the
CompareExchange fails, fall back to the slow path

lambdageek commented:
The interpreter really doesn't like calls.  Inlined `TryEnterInflatestFast`. We're a bit better now, but still slower than baseline for some reason

interp, tiering disabled:

|                       Method |        Job |                                                                         Toolchain |      Mean |    Error |   StdDev | Ratio | RatioSD |
|----------------------------- |----------- |---------------------------------------------------------------------------------- |----------:|---------:|---------:|------:|--------:|
| LockPulseUnlockFlatObjNoHash | Job-HHWOOU | baseline | 212.39 ns | 2.204 ns | 2.062 ns |  1.00 |    0.00 |
| LockPulseUnlockFlatObjNoHash | Job-ZKNSZX |        PR | 179.43 ns | 2.083 ns | 1.846 ns |  0.85 |    0.01 |
|                              |            |                                                                                   |           |          |          |       |         |
|      LockUnlockFlatObjNoHash | Job-HHWOOU | baseline | 110.91 ns | 1.040 ns | 0.973 ns |  1.00 |    0.00 |
|      LockUnlockFlatObjNoHash | Job-ZKNSZX |        PR | 102.46 ns | 0.100 ns | 0.078 ns |  0.93 |    0.01 |
|                              |            |                                                                                   |           |          |          |       |         |
|            GetHashCodeNoLock | Job-HHWOOU | baseline |  18.51 ns | 0.312 ns | 0.291 ns |  1.00 |    0.00 |
|            GetHashCodeNoLock | Job-ZKNSZX |        PR |  38.90 ns | 0.615 ns | 0.575 ns |  2.10 |    0.05 |
|                              |            |                                                                                   |           |          |          |       |         |
|     LockUnlockInflatedObject | Job-HHWOOU | baseline | 113.77 ns | 1.060 ns | 0.992 ns |  1.00 |    0.00 |
|     LockUnlockInflatedObject | Job-ZKNSZX |        PR | 120.09 ns | 0.106 ns | 0.100 ns |  1.06 |    0.01 |

lambdageek committed:
disable the TryGetHashCode fast path helper

It was actually making things slower on the interpreter.

In the JIT it made object.GetHashCode about 50% faster for an object
with a pre-computed hash, but that might mean we need the same kind of
"no-wrapper; then wrapper" fastpath/slowpath that we have for the
Monitor.Enter intrinsic in mini.

lambdageek commented:
I disabled the object.GetHashCode managed path for now.

lambdageek committed:
Fix last exit from inflated monitor

When nest == 1 the caller is responsible for setting the owner to 0 to
indicate that the monitor is unlocked.  (But we leave the nest count
== 1, so that it is correct next time the monitor is entered)

lambdageek commented on src/mono/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.Mono.cs:
```diff
@@ -39,8 +39,13 @@ public static int OffsetToStringData
         private static extern int InternalGetHashCode(object? o);
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static int GetHashCode(object? o)
```
You mean in order to still leave the managed fast path for the JIT since they got a speed boost?  Yea we could do that.

BrzVlad replied:
yes

lambdageek committed:
avoid repeated calls in TryEnterInflatedFast

lambdageek committed:
Combine TryExit and IsEntered into TryExitChecked

avoid duplicated lockword and sync block manipulations

lambdageek committed:
Re-enable managed GetHashCode in JIT; intrinsify in interp

Instead of treating InternalGetHashCode/InternalTryGetHashCode as
intrinsics in the interpreter, intrinsify RuntimeHelpers.GetHashCode
RuntimeHelpers.TryGetHashCode and avoid the managed code entirely when possible

lambdageek committed:
[coop][interp] Fix GC transitions for thunk invoke wrappers

The code that recognizes GC transition icalls in the interpreter was
only assuming that it will see
mono_threads_enter_gc_safe_region_unbalanced /
mono_threads_exit_gc_safe_region_unbalanced

which are used by managed-to-native wrappers.

In most cases for native-to-managed wrappers the marshaller emits
mono_threads_attach_coop / mono_threads_detach_coop and those are
handled elsewhere by setting the needs_thread_attach flag on the
InterpMethod.

However when the mono_marshal_get_thunk_invoke_wrapper API is used, we
emit a thunk invoke wrapper which uses
mono_threads_enter_gc_unsafe_region_unbalanced /
mono_threads_exit_gc_unsafe_region_unbalanced

Recognize those two icalls and set the needs_thread_attach
flag.  (This is slightly more work than necessary -
mono_threads_attach_coop checks if the thread was previously attached
to the runtime and attaches it if it wasn't.  The thunk invoke
wrappers seem to assume the API caller attaches the thread - so we pay
for an extra check.  But on the other hand we don't need to change the
execution-time behavior of the interpreter by reusing existing mechanisms)

lambdageek commented:
This is basically ready now @BrzVlad @vargaz 

Updated `Monitor.Exit` to combine the ownership check and the managed exit code.  There's also a fix for an interpreter/coop GC bug that was exposed by this PR. I pulled the fix out to https://github.com/dotnet/runtime/pull/81773 since it's kind of unrelated.

This is now faster on the JIT in all cases, and faster on (desktop, with tiering disabled) interp in all cases except `GetHashCode` - which I left as an interp intrinsic

**JIT**

|                       Method |        Job |                                                                         Toolchain |      Mean |     Error |    StdDev | Ratio |
|----------------------------- |----------- |---------------------------------------------------------------------------------- |----------:|----------:|----------:|------:|
| LockPulseUnlockFlatObjNoHash | Job-IFELLE | baseline | 41.795 ns | 0.4716 ns | 0.4411 ns |  1.00 |
| LockPulseUnlockFlatObjNoHash | Job-IBVMHK |        PR | 34.217 ns | 0.1579 ns | 0.1477 ns |  0.82 |
|                              |            |                                                                                   |           |           |           |       |
|      LockUnlockFlatObjNoHash | Job-IFELLE | baseline | 21.451 ns | 0.2313 ns | 0.2163 ns |  1.00 |
|      LockUnlockFlatObjNoHash | Job-IBVMHK |        PR | 20.099 ns | 0.1424 ns | 0.1262 ns |  0.94 |
|                              |            |                                                                                   |           |           |           |       |
|            GetHashCodeNoLock | Job-IFELLE | baseline |  9.808 ns | 0.0080 ns | 0.0071 ns |  1.00 |
|            GetHashCodeNoLock | Job-IBVMHK |        PR |  4.736 ns | 0.0486 ns | 0.0431 ns |  0.48 |
|                              |            |                                                                                   |           |           |           |       |
|     LockUnlockInflatedObject | Job-IFELLE | baseline | 24.052 ns | 0.2341 ns | 0.2190 ns |  1.00 |
|     LockUnlockInflatedObject | Job-IBVMHK |        PR | 21.265 ns | 0.1688 ns | 0.1579 ns |  0.88 |

**Interp** (`--interp=-tiering`)


|                       Method |        Job |                                                                         Toolchain |      Mean |    Error |   StdDev | Ratio | RatioSD |
|----------------------------- |----------- |---------------------------------------------------------------------------------- |----------:|---------:|---------:|------:|--------:|
| LockPulseUnlockFlatObjNoHash | Job-QQGPBT | baseline | 214.26 ns | 1.887 ns | 1.765 ns |  1.00 |    0.00 |
| LockPulseUnlockFlatObjNoHash | Job-BZACNQ |        PR | 187.63 ns | 1.189 ns | 0.992 ns |  0.88 |    0.01 |
|                              |            |                                                                                   |           |          |          |       |         |
|      LockUnlockFlatObjNoHash | Job-QQGPBT | baseline | 111.60 ns | 1.624 ns | 1.519 ns |  1.00 |    0.00 |
|      LockUnlockFlatObjNoHash | Job-BZACNQ |        PR | 107.63 ns | 0.122 ns | 0.108 ns |  0.96 |    0.01 |
|                              |            |                                                                                   |           |          |          |       |         |
|            GetHashCodeNoLock | Job-QQGPBT | baseline |  18.48 ns | 0.293 ns | 0.274 ns |  1.00 |    0.00 |
|            GetHashCodeNoLock | Job-BZACNQ |        PR |  17.87 ns | 0.306 ns | 0.286 ns |  0.97 |    0.02 |
|                              |            |                                                                                   |           |          |          |       |         |
|     LockUnlockInflatedObject | Job-QQGPBT | baseline | 115.02 ns | 0.382 ns | 0.319 ns |  1.00 |    0.00 |
|     LockUnlockInflatedObject | Job-BZACNQ |        PR | 110.56 ns | 0.236 ns | 0.221 ns |  0.96 |    0.00 |

lambdageek committed:
add CMPXCHG opcodes to the jiterpreter

For the i64 version, pass the address of the i64 values to the helper
function.

The issue is that since JS doesn't have i64, EMSCRIPTEN_KEEPALIVE
messes with the function's signature and we get import errors in the
JITed wasm module.  Passing by address works around the issue since
addresses are i32.

lambdageek committed:
Assert that CEE_MONO_GET_SP is followed by gc safe enter/exit icalls

Only assert in debug builds

lambdageek committed:
[marshal] Add GCUnsafeTransitionBuilder; use for thunk_invoke_wrapper

Make a "GC Unsafe Transition Builder" - that always calls
mono_threads_attach_coop / mono_threads_detach_coop.

Use it in the native to managed wrappers:
emit_thunk_invoke_wrapper and emit_managed_wrapper

This is a change in behavior for emit_thunk_invoke_wrapper -
previously it directly called
mono_threads_enter_gc_unsafe_region_unbalanced.

That means that compared to the previous behavior, the thunk invoke
wrappers are now a bit more lax: they will be able to be called on
threads that aren't attached to the runtime and they will attach
automatically.

On the other hand existing code will continue to work, with the extra
cost of a check of a thread local var.

Using mono_thread_attach_coop also makes invoke wrappers work
correctly in the interpreter - it special cases
mono_thread_attach_coop but not enter_gc_unsafe.

lambdageek committed:
Revert "[coop][interp] Fix GC transitions for thunk invoke wrappers"

This reverts commit 5b498b8d4e265f8a68fe9453b535e48cf56756ce.

lambdageek commented:
Implemented `cmpxchg.i32` (and `cmpxchg.i64`) jiterpreter support.  (Attn: @kg)

So we get these results on wasm:

|                       Method |        Job |        Toolchain |      Mean |     Error |    StdDev |    Median | Ratio | RatioSD |
|----------------------------- |----------- |----------------- |----------:|----------:|----------:|----------:|------:|--------:|
| LockPulseUnlockFlatObjNoHash |         PR |         Wasm: PR | 340.94 ns |  0.918 ns |  1.374 ns | 340.78 ns |  0.83 |    0.01 |
| LockPulseUnlockFlatObjNoHash | merge-base | Wasm: merge-base | 410.64 ns |  2.336 ns |  3.496 ns | 411.47 ns |  1.00 |    0.00 |
|                              |            |                  |           |           |           |           |       |         |
|      LockUnlockFlatObjNoHash |         PR |         Wasm: PR | 191.05 ns | 11.016 ns | 16.488 ns | 190.31 ns |  0.85 |    0.13 |
|      LockUnlockFlatObjNoHash | merge-base | Wasm: merge-base | 229.22 ns | 11.732 ns | 17.196 ns | 215.65 ns |  1.00 |    0.00 |
|                              |            |                  |           |           |           |           |       |         |
|            GetHashCodeNoLock |         PR |         Wasm: PR |  57.27 ns | 11.809 ns | 16.936 ns |  57.24 ns |  0.79 |    0.23 |
|            GetHashCodeNoLock | merge-base | Wasm: merge-base |  72.78 ns |  0.408 ns |  0.610 ns |  72.60 ns |  1.00 |    0.00 |
|                              |            |                  |           |           |           |           |       |         |
|     LockUnlockInflatedObject |         PR |         Wasm: PR | 200.53 ns |  1.333 ns |  1.995 ns | 199.85 ns |  0.88 |    0.01 |
|     LockUnlockInflatedObject | merge-base | Wasm: merge-base | 227.66 ns |  0.812 ns |  1.215 ns | 227.89 ns |  1.00 |    0.00 |

kg commented on src/mono/wasm/runtime/jiterpreter.ts:
```diff
@@ -1273,6 +1290,25 @@ function generate_wasm_body (
+                append_stloc_tail(builder, getArgU16(ip, 1), WasmOpcode.i32_store);
+                break;
+            case MintOpcode.MINT_MONO_CMPXCHG_I8:
+                // because i64 values can't pass through JS cleanly (c.f getRawCwrap and
```
None of these values enter/exit JS since the implementation of cmpxchg_i64 is C and the trace is pure wasm, so it should be fine to use i64s directly here

kg replied:
Because you're using getRawCwrap there's no JS wrapper involved (see https://github.com/dotnet/runtime/blob/98514de67e76f25b8b7b7965b27a29c2efecb620/src/mono/wasm/runtime/jiterpreter-support.ts#L931 ) - it should be a raw wasm export which is valid to shove into the import table of the trace module and wire everything through, I believe. If it's not possible it *should* be

lambdageek replied:
No, it doesn't work - there's Wasm.instaniate throws an import error. because `EMSCRIPTEN_KEEPALIVE` makes some weird signature for `mono_jiterp_cas_i64` if I pass int64_t arguments by value

lambdageek replied:
It's possible that something else was wrong when I was doing it by value.  I'm not too inclined to investigate it right now to be honest.  We could make an issue to follow-up.

lambdageek commented:
The benchmark perf numbers stayed roughly the same after rebasing.

[desktop](https://github.com/dotnet/runtime/pull/81380#issuecomment-1421284009)

[wasm](https://github.com/dotnet/runtime/pull/81380#issuecomment-1426263595)

kg commented on src/mono/System.Private.CoreLib/src/System/Threading/Monitor.Mono.cs:
```diff
@@ -43,6 +53,7 @@ public static void TryEnter(object obj, ref bool lockTaken)
         public static bool TryEnter(object obj, int millisecondsTimeout)
         {
             bool lockTaken = false;
+
```
Whitespace (not a big problem)

kg commented on src/mono/System.Private.CoreLib/src/System/Threading/ObjectHeader.Mono.cs:
```diff
@@ -0,0 +1,451 @@
+    // stack that stores a pointer to the object, thus pinning the object.
+    private unsafe ref struct ObjectHeaderOnStack
+    {
+        private Header** _header;
```
If this is a Header** doesn't that mean there won't be a visible pointer to the object on the stack? Instead, it's a pointer-to-pointer, right?

lambdageek replied:
`_header` is a pointer to pointer.  Which means the thing that it points to is on the stack (well... the thing it points to is in memory - but we happen to know it's on the stack) - that's the one we want the GC to notice.

lambdageek committed:
fix whitespace

lambdageek committed:
Merge remote-tracking branch 'origin/main' into feature-lockword-in-managed

lambdageek committed:
more whitespace

lambdageek committed:
[jiterp] Allow null obj in hashcode intrinsics

The underlying C functions are null-friendly, and the managed
intrinsics are allowed to C null inputs.  The assert is unnecessary.
